desc/objectives/_stability.py

```diff
@@ -367,3 +369,278 @@ def compute(self, params, constants=None):
+
+    where λ is the negative squared growth rate for each field line (such that λ>0 is
+    unstable), λ₀ is a cutoff, and w₀ and w₁ are weights.
+
+    Parameters
+    ----------
+    eq : Equilibrium
+        Equilibrium that will be optimized to satisfy the Objective.
+    target : {float, ndarray}, optional
+        Target value(s) of the objective. Only used if bounds is None.
+        Must be broadcastable to Objective.dim_f. Default is ``target=0``
+    bounds : tuple of {float, ndarray}, optional
+        Lower and upper bounds on the objective. Overrides target.
+        Both bounds must be broadcastable to to Objective.dim_f. Default is ``target=0``
+    weight : {float, ndarray}, optional
+        Weighting to apply to the Objective, relative to other Objectives.
+        Must be broadcastable to to Objective.dim_f
+    normalize : bool, optional
+        Whether to compute the error in physical units or non-dimensionalize.
+        Not used since the growth rate is always normalized.
+    normalize_target : bool, optional
+        Whether target and bounds should be normalized before comparing to computed
+        values. If `normalize` is `True` and the target is in physical units,
+        this should also be set to True. Not used since the growth rate is always
+        normalized.
+    loss_function : {None, 'mean', 'min', 'max'}, optional
+        Loss function to apply to the objective values once computed. This loss function
+        is called on the raw compute value, before any shifting, scaling, or
+        normalization. Has no effect for this objective.
+    deriv_mode : {"auto", "fwd", "rev"}
+        Specify how to compute jacobian matrix, either forward mode or reverse mode AD.
+        "auto" selects forward or reverse mode based on the size of the input and output
+        of the objective. Has no effect on self.grad or self.hess which always use
+        reverse mode and forward over reverse mode respectively.
+    rho : float
+        Flux surface to optimize on. To optimize over multiple surfaces, use multiple
+        objectives each with a single rho value.
+    alpha : float, ndarray
+        Field line labels to optimize. Values should be in [0, 2pi). Default is alpha=0
+        for axisymmetric equilibria, or 8 field lines linearly spaced in [0, pi] for
+        non-axisymmetric cases.
+    nturns : int
+        Number of toroidal transits of a field line to consider. Field line
+        will run from -π*nturns to π*nturns. Default 3.
+    nzetaperturn : int
+        Number of points along the field line per toroidal transit. Total number of
+        points is ``nturns*nzetaperturn``. Default 100.
+    zeta0 : array-like
+        Points of vanishing integrated local shear to scan over.
+        Default 15 points in [-π/2,π/2]
+    lam0 : float
+        Threshold for penalizing growth rates in metric above.
+    w0, w1 : float
+        Weights for sum and max terms in metric above.
```

Comment:
How are these different from the `target` and `weight` or are they redundant? Are they only necessary for benchmarking to COBRAVMEC, or do they actually affect the physics?

Reply:
It's completely independent from COBRAVMEC. We don't even use it when comparing our results with COBRAVMEC.
These weights are for the optimization part, not the valuation part.

Reply:
In theory the same effect could be obtained with regular weight and bounds but that seems to cause some memory issues due to the extra residuals, see my previous comment about it

Reply:
@f0uriest Can you link to your previous comment? I don't see it

What extra residuals? This is a scalar objective, so I don't understand how our usual objective target/weight could be causing memory issues.

Reply:
What I mean is that this objective is using a custom loss function of the form
```
f = w₀ sum(ReLU(λ-λ₀)) + w₁ max(ReLU(λ-λ₀))
```

which requires the w0,w1,lambda0 parameters, and the sum/max are taken over all the growth rates on the surface (ie over alpha and zeta0). In theory if we set `bounds=(-np.inf, lambda0)` and `weight=w0` this would be the same as the first term above when reduced to a scalar, but the actual optimizer would see a vector of residuals and differentiating all of them (rather than just the sum) leads to OOM errors, hence using the custom loss function to reduce things down to a scalar _inside_ the `compute` method rather than returning all the residuals and doing the reduction implicitly later (ie when the optimizer does `f.T@f`)

Reply:
https://github.com/PlasmaControl/DESC/pull/1170#discussion_r1739060067

Reply:
Ah ok I didn't realize this was really a vector-valued function with multiple growth rates. Don't we usually only care about the most unstable mode? Why can't we simply return the maximum growth rate? That's what TERPSICHORE does. Is it just that empirically we notice that minimizing the sum of all the unstable modes happens to work better?

Reply:
Yes, the sum of growth rates at multiple points works better. This is what we are doing

![Screenshot from 2024-09-10 00-10-11](https://github.com/user-attachments/assets/91f61c15-abcd-49d7-87d8-2b3b17d61b6b)

This is like the lambda values on different points on a flux surface. Sometimes, when you just try to supress the maximum value, it rises at a different location. Suppressing all positive lambda values pushes all positive values together until everything is stable. It's certainly more robus.

Reply:
We are still taking the most unstable mode but at multiple locations (flux tubes) on a flux surface. Please read the current paper that I shared with you or my older adjoint ballooning paper to understand this better.